text/0000-global-packages.md

```diff
@@ -0,0 +1,151 @@
+
+The core of this approach is to take advantage of the existing behavior of package manager global installs like `npm i -g` and `yarn global add`. Those tools already provide the expected interoperability between global packages that we want and users have intuition around how they work. We can insert some customization before the global install runs to point it at a custom directory that Volta controls, so we can keep track of everything. Additionally, we can do record-keeping after the global install finishes to ensure that we keep track of which tools and binaries are installed.
+
+Finally, to avoid forcing the user to re-install every global package whenever they switch Node versions, we can use our manifest of installed packages to automatically re-install them whenever the user changes their default Node version. That way they get the benefits of global installs without the added hassles.
+
+## Sandbox Per Node Major Version
+
+The major issue preventing us from having a single persistent store of global packages is that Node major versions can represent breaking changes that cause a package to stop working. This is most notable with native modules, which are compiled against a specific Node Modules API version and need to be recompiled to work against a different Node. Since these types of changes are primarily limited to major version bumps in Node, we can keep a separate cache of global packages for each major version. This will be stored within the Volta directory, most likely under `tools/image/packages/<node version>`.
+
+Within a given Node Major bucket, we expect packages to continue to work without needing to be re-installed, so we can avoid needing to have a separate cache for every individual Node version. This also allows us to reduce the amount of time we spend re-installing packages on Node version switches.
+
+## Interception of Global Installs
+
+When a user runs `npm i -g <package>` or `yarn global add <package>`, instead of throwing a hard error as we currently do, we can make 2 small changes to the workflow but otherwise allow the command to run as expected:
+
+1. Use the `prefix` and `global-folder` options to redirect the install directory to the Volta-managed location. Package managers already provide support for changing the global install directory, so we can leverage that to make sure we always install into the appropriate Node major version cache. We should set these (if possible) using environment variables, so that they can still be overridden by command-line switches from the user in cases where they want to have control over the directories.
+
+2. After the install completes, take inventory of the packages that changed (whether installed, upgraded, or uninstalled) to update an internal manifest of all packages the user has globally installed.
+
+Additionally, if a user calls `volta install <package>`, we can perform the same steps and shell out to `npm i -g <package>` as if the user had typed that directly.
```

Comment:
I wonder if we should just deprecate this during this RFC. We have already overloaded the `install` command npm global packages + "primitive" executables and this RFC seems like the appropriate time to deprecate this particular subcase.

Reply:
I've gone back and forth a bit on this. On the one hand, reducing the overload of `volta install` is nice, but on the other, the ability to quickly spin up an environment with `volta install node yarn ember-cli` is _really_ nice.

Ultimately, I think that the discussion of changing the semantics of `volta install` is separate from this (though potentially dependent on this change), so it should be its own RFC. As long as `volta install` keeps the current behavior, I think it's useful to keep the package support around (even if it's mostly a pass-through to `npm i -g <package>`). When / if we decide to change around the commands, that's a better point to talk about deprecating support for packages directly through Volta.